prog/github/repository_nexus.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require_relative "../../lib/util"
+
+class Prog::Github::RepositoryNexus < Prog::Base
+  subject_is :github_repository
+
+  semaphore :destroy
+
+  def self.assemble(installation, name)
+    DB.transaction do
+      repository = GithubRepository.new_with_id(installation_id: installation.id, name: name)
+      repository.skip_auto_validations(:unique) do
+        repository.insert_conflict(target: [:installation_id, :name], update: {last_job_at: Time.now}).save_changes
+      end
+      Strand.new(prog: "Github::RepositoryNexus", label: "wait") { _1.id = repository.id }
+        .insert_conflict(target: :id).save_changes
+    end
+  end
+
+  def client
+    @client ||= Github.installation_client(github_repository.installation.installation_id).tap { _1.auto_paginate = true }
+  end
+
+  def check_queued_jobs
+    queued_runs = client.repository_workflow_runs(github_repository.name, {status: "queued"})[:workflow_runs]
+    Clog.emit("polled queued runs") { {polled_queued_runs: {repository_name: github_repository.name, count: queued_runs.count}} }
+    queued_labels = {}
+    queued_runs.each do |run|
+      jobs = client.workflow_run_attempt_jobs(github_repository.name, run[:id], run[:run_attempt])[:jobs]
+
+      jobs.each do |job|
+        next if job[:status] != "queued"
+        next unless (label = job[:labels].find { Github.runner_labels.key?(_1) })
+        queued_labels[label] ||= 0
+        queued_labels[label] += 1
+      end
+
+      Clog.emit("polled queued jobs") { {polled_queued_jobs: {repository_name: github_repository.name, count: queued_labels.values.sum}} }
+    end
+
+    queued_labels.each do |label, count|
+      idle_runner_count = github_repository.runners_dataset.where(label: label, workflow_job: nil).count
+      next if (required_runner_count = count - idle_runner_count) && required_runner_count <= 0
+
+      Clog.emit("extra runner needed") { {needed_extra_runner: {repository_name: github_repository.name, label: label, count: required_runner_count}} }
+
+      required_runner_count.times do
+        Prog::Vm::GithubRunner.assemble(
+          github_repository.installation,
+          repository_name: github_repository.name,
+          label: label
+        )
+      end
+    end
+  rescue Octokit::NotFound
+    github_repository.incr_destroy
```

Comment:
we can simply `hop`? Even if we don't is it really OK to call `nap` in a method that is not `label`? I would prefer it to have a return type and handle the state transition command in a `label` function. That way, it's more clear what the state transition is in which case.

Reply:
We avoid calling `hop_destroy` directly in our codebase to ensure we don't bypass `before_run`. However, I agree that using `nap` in a method can make the code somewhat unclear. I'll work on refactoring it.